eio/tls_eio.ml

```diff
@@ -5,198 +5,143 @@ exception Tls_failure of Tls.Engine.failure
     recv_buf       : Cstruct.t ;
+    mutex          : Eio.Mutex.t;
   }
 
-  let read_t t cs =
-    try Flow.read t.flow cs
-    with
-    | End_of_file as ex ->
-      t.state <- `Eof;
+  let write_t t cs = Flow.copy (Flow.cstruct_source [cs]) t.flow
+
+  (* A TLS 'read' requires both reading and writing to an underlying
+     flow object - t.flow; therefore we use mutex to maintain the
+     integrity of 'read' from multiple fibres doing interleaving
+     read/write operations on the same underlying flow object. *)
+  let read_t t =
+    try
+      Eio.Mutex.lock t.mutex ;
+      let got = Flow.single_read t.flow t.recv_buf in
+      let data = Cstruct.sub t.recv_buf 0 got in
+      match Tls.Engine.handle_tls t.tls data with
+      | Ok ( state, `Response resp, `Data application_data ) ->
+        begin match state with
+          | `Ok tls ->
+            t.tls <- tls ;
+            Option.iter (write_t t) resp ;
+            Eio.Mutex.unlock t.mutex ;
+            application_data
+          | `Eof -> raise End_of_file ;
+          | `Alert a -> raise (Tls_alert a)
+        end
+      | Error (failure, `Response resp) ->
+        write_t t resp ;
+        raise (Tls_failure failure)
+    with 
+    | ( End_of_file | Tls_alert _ | Tls_failure _ ) as ex ->
+      Eio.Mutex.unlock t.mutex ;
       raise ex
-    | exn ->
-      (match t.state with
-       | `Error _ | `Eof -> ()
-       | `Active _ -> t.state <- `Error exn) ;
-      raise exn
-
-  let write_t t cs =
-    try Flow.copy (Flow.cstruct_source [cs]) t.flow
-    with exn ->
-      (match t.state with
-       | `Error _ | `Eof -> ()
-       | `Active _ -> t.state <- `Error exn) ;
-      raise exn
-
-  let try_write_t t cs =
-    try write_t t cs
-    with _ -> Eio.Fiber.check ()      (* Error is in [t.state] *)
-
-  let rec read_react t =
-
-    let handle tls buf =
-      match Tls.Engine.handle_tls tls buf with
-      | Ok (state', `Response resp, `Data data) ->
-          let state' = match state' with
-            | `Ok tls  -> `Active tls
-            | `Eof     -> `Eof
-            | `Alert a -> `Error (Tls_alert a)
-          in
-          t.state <- state' ;
-          Option.iter (try_write_t t) resp;
-          data
-
-      | Error (alert, `Response resp) ->
-          t.state <- `Error (Tls_failure alert) ;
-          write_t t resp; read_react t
-    in
 
-    match t.state with
-    | `Error e  -> raise e
-    | `Eof      -> raise End_of_file
-    | `Active _ ->
-        let n = read_t t t.recv_buf in
-        match (t.state, n) with
-        | (`Active tls, n) -> handle tls (Cstruct.sub t.recv_buf 0 n)
-        | (`Error e, _)    -> raise e
-        | (`Eof, _)        -> raise End_of_file
-
-  let rec read t buf =
-
-    let writeout res =
-      let open Cstruct in
-      let rlen = length res in
-      let n    = min (length buf) rlen in
-      blit res 0 buf 0 n ;
+  let rec read t flow_buf : int =
+
+    let write_application_data data =
```

Comment:
from the name of this function, this sounds like "only application data will ever be written here", but AFAICT there could as well be handshake data that is being written. how did you choose the name?